File: frontend/src/types/Task.ts
/**
 * Task-related TypeScript types
 */

export enum TaskStatus {
  PENDING = "pending",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
  BLOCKED = "blocked",
  CANCELLED = "cancelled"
}

export enum TaskSource {
  EMAIL = "email",
  MEETING = "meeting",
  MANUAL = "manual",
  CALENDAR = "calendar"
}

export interface Task {
  id: string;
  title: string;
  description?: string;
  assigned_to?: string;
  created_by: string;
  urgency: number; // 1-5
  deadline?: string; // ISO datetime
  estimated_effort?: number; // Hours
  status: TaskStatus;
  priority_score?: number; // 0-1
  dependencies: string[]; // Task IDs
  source: TaskSource;
  source_metadata?: Record<string, any>;
  completed_at?: string;
  actual_effort?: number;
  created_at: string;
  updated_at: string;
}

export interface TaskCreate {
  title: string;
  description?: string;
  assigned_to?: string;
  created_by: string;
  urgency?: number;
  deadline?: string;
  estimated_effort?: number;
  source?: TaskSource;
  source_metadata?: Record<string, any>;
  dependencies?: string[];
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  urgency?: number;
  deadline?: string;
  estimated_effort?: number;
  status?: TaskStatus;
  assigned_to?: string;
  actual_effort?: number;
}

export interface TaskWithDetails extends Task {
  assigned_employee_name?: string;
  creator_name: string;
}

export interface TaskCandidate {
  title: string;
  description?: string;
  urgency: number;
  estimated_effort?: number;
  deadline?: string;
  confidence: number; // 0-1
}

export interface ScheduledTask {
  task_id: string;
  task_title: string;
  priority_score: number;
  suggested_start: string;
  suggested_end: string;
  urgency: number;
  deadline?: string;
}

File: backend/app/models/task.py
"""Task database model"""
from sqlalchemy import Column, String, Text, Integer, Float, DateTime, ForeignKey, Enum, JSON, ARRAY
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import enum
from .base import BaseModel


class TaskStatus(str, enum.Enum):
    """Task status enumeration"""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    BLOCKED = "blocked"
    CANCELLED = "cancelled"


class TaskSource(str, enum.Enum):
    """Task source enumeration"""
    EMAIL = "email"
    MEETING = "meeting"
    MANUAL = "manual"
    CALENDAR = "calendar"


class Task(BaseModel):
    """Task model with automatic prioritization"""
    __tablename__ = "tasks"

    title = Column(String(500), nullable=False)
    description = Column(Text, nullable=True)

    # Assignment
    assigned_to = Column(UUID(as_uuid=True), ForeignKey("employees.id"), nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("employees.id"), nullable=False)

    # Priority and urgency
    urgency = Column(Integer, nullable=False, default=3)  # 1-5 scale (manager input)
    deadline = Column(DateTime, nullable=True)
    estimated_effort = Column(Float, nullable=True)  # Hours

    # Status
    status = Column(Enum(TaskStatus), nullable=False, default=TaskStatus.PENDING)

    # AI-calculated priority score
    priority_score = Column(Float, nullable=True)  # 0-1 scale

    # Dependencies (array of task UUIDs)
    dependencies = Column(JSON, default=list)  # List of UUID strings

    # Source information
    source = Column(Enum(TaskSource), nullable=False, default=TaskSource.MANUAL)
    source_metadata = Column(JSON, nullable=True)  # Email ID, meeting link, etc.

    # Completion tracking
    completed_at = Column(DateTime, nullable=True)
    actual_effort = Column(Float, nullable=True)  # Actual hours spent

    # Relationships
    assigned_employee = relationship("Employee", back_populates="tasks", foreign_keys=[assigned_to])
    creator = relationship("Employee", back_populates="created_tasks", foreign_keys=[created_by])

    def __repr__(self):
        return f"<Task(id={self.id}, title={self.title}, status={self.status}, priority={self.priority_score})>"
